src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -4,67 +4,186 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using Microsoft.Extensions.ObjectPool;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Razor.PooledObjects;
 
 /// <summary>
 ///  Wraps a pooled <see cref="ImmutableArray{T}.Builder"/> but doesn't allocate it until
 ///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
 ///  to the pool.
+///
+///  There is significant effort to avoid retrieving the <see cref="ImmutableArray{T}.Builder"/>.
+///  For very small arrays of length 4 or less, the elements will be stored on the stack. If the array
+///  grows larger than 4 elements, a builder will be employed.
 /// </summary>
-internal ref partial struct PooledArrayBuilder<T>
+internal partial struct PooledArrayBuilder<T> : IDisposable
 {
-    private readonly ObjectPool<ImmutableArray<T>.Builder> _pool;
-    private readonly int? _capacity;
+    public static readonly PooledArrayBuilder<T> Empty = default;
+
+    /// <summary>
+    ///  The number of items that can be stored inline.
+    /// </summary>
+    private const int InlineCapacity = 4;
+
+    /// <summary>
+    ///  A builder used as storage when the number of items exceeds 4 elements.
```

Comment:
This isn't _quite_ right. For example, this code will still have a `_builder`:

```cs
var builder = new PooledArrayBuilder<int>();
builder.AddRange([1, 2, 3, 4, 5]);
builder.RemoveAt(0); // We now have 4 elements, but still access through _builder always
```

I think it's worth a comment to call this behavior, and specify that `builder` must _always_ be prefered for access if non-null, even if the count <= InlineCapacity.